tests/common/hashkv_test.go

```diff
@@ -0,0 +1,248 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"go.etcd.io/etcd/tests/v3/framework/config"
+	intf "go.etcd.io/etcd/tests/v3/framework/interfaces"
+	"go.etcd.io/etcd/tests/v3/framework/testutils"
+)
+
+// TestVerifyHashKVAfterCompact tests that HashKV is consistent across all members after a physical compaction.
+// It tests both cases where the compaction is on a tombstone revision and where it is not.
+func TestVerifyHashKVAfterCompact(t *testing.T) {
+	testRunner.BeforeTest(t)
+	for _, keys := range [][]string{
+		{"key0"},
+		{"key0", "key1"},
+	} {
+		for _, compactedOnTombstoneRev := range []bool{false, true} {
+			t.Run(fmt.Sprintf("Keys=%v/CompactedOnTombstone=%v", keys, compactedOnTombstoneRev), func(t *testing.T) {
+				for _, tc := range clusterTestCases() {
+					t.Run(tc.name, func(t *testing.T) {
+						if tc.config.ClusterSize < 2 {
+							t.Skip("Skipping test for single-member cluster")
+						}
+						ctx, cancel := context.WithTimeout(t.Context(), 30*time.Second)
+						defer cancel()
+
+						clus := testRunner.NewCluster(ctx, t, config.WithClusterConfig(tc.config))
+						defer clus.Close()
+
+						cc := testutils.MustClient(clus.Client())
+						tombstoneRevs, latestRev := populateDataForHashKV(t, cc, keys)
+
+						compactedOnRev := tombstoneRevs[0]
+
+						// If compaction revision isn't a tombstone, select a revision in the middle of two tombstones.
+						if !compactedOnTombstoneRev {
+							require.Greater(t, len(tombstoneRevs), 1)
+							compactedOnRev = (tombstoneRevs[0] + tombstoneRevs[1]) / 2
+							require.Greater(t, compactedOnRev, tombstoneRevs[0])
+							require.Greater(t, tombstoneRevs[1], compactedOnRev)
+						}
+
+						_, err := cc.Compact(ctx, compactedOnRev, config.CompactOption{Physical: true})
+						require.NoError(t, err)
+
+						for rev := compactedOnRev; rev <= latestRev; rev++ {
+							verifyConsistentHashKVAcrossAllMembers(t, cc, rev)
+						}
+					})
+				}
+			})
+		}
+	}
+}
+
+// TestVerifyHashKVAfterTwoCompactionsOnTombstone tests that HashKV is consistent
+// across all members after two physical compactions on tombstone revisions.
+func TestVerifyHashKVAfterTwoCompactionsOnTombstone(t *testing.T) {
+	testRunner.BeforeTest(t)
+	for _, tc := range clusterTestCases() {
+		t.Run(tc.name, func(t *testing.T) {
+			if tc.config.ClusterSize < 2 {
+				t.Skip("Skipping test for single-member cluster")
+			}
```

Comment:
The test scope is across cluster members, so it can be skipped in the single-member case.